### ISSUE TITLE
cli/sql: `\!` and `\|` accumulate data in RAM and cause the client to crash on a large input

### ISSUE THREAD

**Issue by jlauro** [cockroachdb/cockroach#15645]
tldr: the `\!` (run command) and `\|` (pipe command outputs)  accumulate the output of the command in RAM before using it. **If the command has a lot of output, the client can crash.**

This is not necessary: on unix it is possible to consume the output of a commoand in a streaming fashion, e.g. by using a pipe created with `os.Pipe()`.

This issue becomes critical in light of the server running initialization SQL scripts (see  #54771): we don't want the server to crash because of a large input file.

After the refactoring in #54796 and #54741 , the shell has become properly recursive. Here are the missing bits of work to fix this:

1) for `\!`, connect the command to the shell's `os.Stdout`. There's no need to process the output so the command can write to the same stdout as the shell.

2) for `\|`, call `os.Pipe`, then:

   1. `cmd.Start()`
   2. `runRecursive()`
   3. `cmd.Wait()`

(with proper error handling, etc)

----

Original text from the issue poster

Config info:
crdb3a.ces.cvnt.net:/data/crdb/logs# grep -F '[config]' cockroach.log 
I170503 00:14:31.063814 89 util/log/clog.go:887  [config] file created at: 2017/05/03 00:14:31
I170503 00:14:31.063814 89 util/log/clog.go:887  [config] running on machine: crdb3a
I170503 00:14:31.063814 89 util/log/clog.go:887  [config] binary: CockroachDB CCL v1.0-rc.1-dirty (linux amd64, built 2017/05/01 18:33:34, go1.8.1)
I170503 00:14:31.063814 89 util/log/clog.go:887  [config] arguments: [cockroach start --store=path=/data/crdb --locality=datacenter=ohq]
(also previous beta)

Observed behavior:
Using \| cat ... works on small files but fails on large files.  See below on test import file (17mb) (and retried with redirect from command line which works)

crdb3a.ces.cvnt.net:/root/e/employees_db# cockroach sql    
# Welcome to the cockroach SQL interface.
# All statements must be terminated by a semicolon.
# To exit: CTRL + D.
root@:26257/> \| cat load_employees.dump
driver: bad connection
connection lost; opening new connection and resetting session parameters...
root@:26257/> ^D
Error: driver: bad connection
Failed running "sql"
crdb3a.ces.cvnt.net:/root/e/employees_db# cockroach sql <load_employees.dump 
SET
INSERT 500
INSERT 500
INSERT 500
INSERT 500
...
Note: Server does not have a crash or otherwise have problem on the disconnect.  Appears to be client side bug with large file.

Jira issue: CRDB-6085

**Comment by petermattis** [comment]
@jlauro Thanks for the report. The "bad connection" messages make me suspect something is going wrong on the server. Can you share your server logs? It is possible the server crashed. We also recently introduced a limit on the size of a transaction. Perhaps you're encountering that limit (though I would expect a different error than "bad connection").

Even better would be if you can share `load_employees.dump` so that we can reproduce the problem locally.

**Comment by jlauro** [comment]
I verified nothing in the server logs at the time.  Waited for a pause in activity, and did the \| cat by hand and made certain logs did not change until after the command had returned and failed.  It returns quickly...

It's a test database and the original data was made for mysql, but I tweaked the table creation to be compatible with cockroachdb and the inserts to only do 500 rows per line, or under 32k, so they would be reasonable, unlike the original that had 1mb rows... Limits on this site restricted me from including all the test files.
[testdata.zip](https://github.com/cockroachdb/cockroach/files/975224/testdata.zip)

**Comment by petermattis** [comment]
@jlauro Thanks for the test data. I'll take a look as soon as I can download it (over my currently very poor internet connection).

**Comment by petermattis** [comment]
I've reproduced the "bad connection" error. The server definitely didn't crash.

**Comment by petermattis** [comment]
When loading the employee data, the server is closing the connection with this error:

```
E170504 08:58:41.452828 495 server/server.go:625  [n1,client=[::1]:53703] message size 17439983 out of bounds (0..16777216)
```

`17439983` is exactly the size of `load_employees.dump`. @knz I believe you added the support for `\| cat` to the sql shell. Is it sending the contents of the file as a single operation?

PS @tamird The 16 MB limit here is from `sql/pgwire/encoding.go:maxMessageSize`. Looks like we've had that limit forever. Do you know where that value came from? I don't think raising it is necessarily wise, just curious.

**Comment by knz** [comment]
Yes the file is loaded first in memory in the client and then sent as a single query.

**Comment by knz** [comment]
We could possibly change pgwire to return a query error instead of closing the connection.

**Comment by knz** [comment]
@jlauro To summarize, when you use `|\ cat...` this sends the entire dump file as a single query to the server. the server has an upper bound of 16MB on the size of a packet it can accept from any client; if a packet exceeds this size, the connection is closed by the server. This causes the error you observe.

We hadn't envisioned that `\|` would be used to load a lot of data. Perhaps this needs to be changed so that `\|` behaves more like `... | cockroach sql`.

**Comment by petermattis** [comment]
It is somewhat surprising that `\| cat file` behaves differently than `cat file | cockroach sql`. The latter will breaks the input into statements and sends individual statements.

**Comment by knz** [comment]
@petermattis it's an artifact of how the shell is architectured. Currently the entire read-eval-print loop assumes that built-in commands only have 1 "line" of effect. Changing this behavior isn't hard but would change the code in that area significantly.

**Comment by petermattis** [comment]
@knz Got it.

**Comment by mberhault** [comment]
@knz: from your comment above, I take it this would be too disruptive to do for 1.1 now?

**Comment by knz** [comment]
No, it would not be disruptive actually.

**Comment by knz** [comment]
The way to do this:

1. make `isInteractive` as used throughout `cli/sql.go` and `sql_util.go` a member of `cliState`, instead of a global variable
1. make a sub-structs for `cliState`: `perFileCliState`. make `isInteractive` part of `perFileCliState` as well as `buf`, `ins`, `atEOF`, `lastInputLine`, `partialLines`, `partialStmtsLen`, `concatLines`. The other fields stay in `cliState`
1. When encountering a pipe command, i.e. in `pipeSysCmd` one would:
    - spawn a subprocess with pipes (`exec.Command`)
    - create a new `perFileCliState` with `isInteractive` set to false, `buf` bound to the sub-process'es stdout stream, the other fields initialized to default
    - recurse into `runInteractive` using the new `perFileCliState`
    - when the recursion ends, restore the original `perFileCliState`.

**Comment by petermattis** [comment]
This may not be disruptive, but is it a priority given that there is a workaround that has been documented?

**Comment by knz** [comment]
Perhaps not. What about making it a starter project?

**Comment by petermattis** [comment]
SGTM

**Comment by sploiselle** [comment]
@knz Can we get a docs description this issue as a Known Limitation for 2.1?

**Comment by knz** [comment]
cc @RaduBerinde as you were asking about this yesterday
